File: typescript/src/verification/open-net/predicate.ts
/**
 * @module open-net/predicate
 *
 * The contract's guarantee at one quiescent marking ([VER-022]): which clauses it breaks and
 * which places it strands. Stranding is the [VER-014] rest set every `DeadlockFree` route
 * reads, so the SMT route can ask for it with `deadlockFree()`.
 */
import type { Place } from '../../core/place.js';
import { compareCodePoints } from '../../core/internal/code-point-order.js';
import type { MarkingState } from '../marking-state.js';
import { tokensAcross, countViolation } from '../graph-decision.js';
import { strandedPlaces, type ConditionalSinks } from '../rest-set.js';
import { countAcross } from '../smt-property.js';
import type { ClosedNet } from './closure.js';
import type { CountClause, OpenNetContract } from './contract.js';

/** One thing a quiescent marking breaks. */
export type Finding =
  | { readonly kind: 'clause'; readonly clause: CountClause; readonly count: number; readonly bound: 'lower' | 'upper' }
  | { readonly kind: 'stranded'; readonly place: string; readonly count: number };

/** The contract's rest set in [VER-014] form. */
export interface RestDeclaration {
  readonly sinks: ReadonlySet<Place<any>>;
  readonly conditional: readonly ConditionalSinks[];
}

/**
 * Clause places, rest places and the environment's own places as sinks; each designed
 * terminal as a conditional sink.
 */
export function restDeclarationOf(contract: OpenNetContract, closed: ClosedNet): RestDeclaration {
  const sinks = new Map<string, Place<any>>();
  for (const c of contract.clauses) for (const p of c.places) if (!sinks.has(p.name)) sinks.set(p.name, p);
  for (const p of contract.rest) if (!sinks.has(p.name)) sinks.set(p.name, p);
  for (const p of closed.environmentPlaces) if (!sinks.has(p.name)) sinks.set(p.name, p);
  return {
    sinks: new Set(sinks.values()),
    conditional: contract.terminals.map(t => ({ marker: t.marker, places: new Set(t.excused) })),
  };
}

/** The clause lower bounds' waivers, for both routes: every designed terminal's marker ([VER-002]). */
export function waiverMarkers(contract: OpenNetContract): Place<any>[] {
  return contract.terminals.map(t => t.marker);
}

/**
 * The places `m` strands, in code-point order of their names ([VER-013]). Both routes
 * attribute a stranding through this: marked **and** unexcused, with the [VER-014]
 * widening, so neither reports a stranding the other proves impossible.
 */
export function strandedNames(m: MarkingState, rest: RestDeclaration): Place<any>[] {
  return strandedPlaces(m, rest.sinks, rest.conditional).sort((a, b) => compareCodePoints(a.name, b.name));
}

/**
 * What the quiescent marking `m` breaks: clauses in contract order, then stranded places by
 * name. Empty when `m` meets the contract.
 */
export function quiescenceFindings(m: MarkingState, contract: OpenNetContract, rest: RestDeclaration): Finding[] {
  const findings: Finding[] = [];
  // Each clause is a QuiescentCount waived by every terminal marker ([VER-002]).
  const markers = waiverMarkers(contract);
  for (const clause of contract.clauses) {
    const bound = countViolation(m, clause.places, clause.min, clause.max, markers);
    if (bound !== null) findings.push({ kind: 'clause', clause, count: tokensAcross(m, clause.places), bound });
  }
  for (const place of strandedNames(m, rest)) {
    findings.push({ kind: 'stranded', place: place.name, count: m.tokens(place) });
  }
  return findings;
}

/** The finding's subject: its clause's name or its place's name. */
export function subjectOf(f: Finding): string {
  return f.kind === 'clause' ? f.clause.name : f.place;
}

/** The finding in words. */
export function describeFinding(f: Finding): string {
  if (f.kind === 'stranded') {
    return `${f.place} holds ${f.count} at quiescence, and nothing in the contract lets a token rest there`;
  }
  const { clause } = f;
  const expected = countAcross(clause.min, clause.max, clause.places);
  return f.bound === 'upper'
    ? `${expected} at quiescence, found ${f.count} (an upper bound holds under a terminal too)`
    : `${expected} at quiescence, found ${f.count}`;
}

File: rust/libpetri-verification/src/open_net/predicate.rs
//! The contract's guarantee at one quiescent marking ([VER-022]): which clauses it breaks
//! and which places it strands.
//!
//! The stranding half is the rest set of [VER-014], read through the same module every
//! `DeadlockFree` route reads ([`crate::rest_set`]). The clause places, the rest places and
//! the environment's own places are sinks, and each designed terminal is a conditional
//! sink. That is what lets the SMT route ask for it with `DeadlockFree` and mean the same
//! thing.
//!
//! Everything both routes judge by is derived here once — the rest declaration, the waiver
//! markers and the stranding attribution — and neither route rebuilds any of it: when two
//! call sites each assembled their own copy, they drifted, and one route reported a
//! stranding the other proved impossible.

use std::collections::HashSet;

use crate::graph_decision::{CountBound, count_violation, tokens_across};
use crate::marking_state::MarkingState;
use crate::property::count_across;
use crate::rest_set::{ConditionalSinks, stranded_places};

use super::closure::ClosedNet;
use super::contract::{CountClause, OpenNetContract};

/// The contract's rest set in [VER-014] form.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(super) struct RestDeclaration {
    /// Clause places, then rest places, then the environment's own places, each once.
    pub sinks: Vec<String>,
    pub conditional: Vec<ConditionalSinks>,
}

/// Clause places, rest places and the environment's own places as sinks; each designed
/// terminal as a conditional sink.
pub(super) fn rest_declaration_of(contract: &OpenNetContract, closed: &ClosedNet) -> RestDeclaration {
    let mut seen: HashSet<&str> = HashSet::new();
    let mut sinks: Vec<String> = Vec::new();
    let named = contract
        .clauses()
        .iter()
        .flat_map(|c| c.places.iter())
        .chain(contract.rest().iter())
        .chain(closed.environment_places.iter());
    for p in named {
        if seen.insert(p.as_str()) {
            sinks.push(p.clone());
        }
    }
    let conditional = contract
        .terminals()
        .iter()
        .map(|t| ConditionalSinks { marker: t.marker.clone(), places: t.excused.clone() })
        .collect();
    RestDeclaration { sinks, conditional }
}

/// The clause lower bounds' waivers: every designed terminal's marker ([VER-002]).
///
/// Both routes must waive by the same set in the same order — the graph route reads it
/// through [`count_violation`] and the encoder through its index order — so it is derived
/// here once rather than at each call site.
pub(super) fn waiver_markers(contract: &OpenNetContract) -> Vec<String> {
    contract.terminals().iter().map(|t| t.marker.clone()).collect()
}

/// The places `m` strands, by name, in code-point order.
///
/// This is the predicate both routes must attribute a stranding with, and the reason it
/// lives here rather than at each call site: it asks whether a place is marked **and
/// unexcused**, applying the [VER-014] widening that a token on a place excused by a marked
/// terminal marker is designed residue. A caller that asks only "is this place marked?"
/// reports a stranding the other route proves cannot happen.
pub(super) fn stranded_names(m: &MarkingState, rest: &RestDeclaration) -> Vec<String> {
    stranded_places(m, &rest.sinks, &rest.conditional)
}

/// One thing a quiescent marking breaks.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(super) enum Finding<'c> {
    Clause { clause: &'c CountClause, count: usize, bound: CountBound },
    Stranded { place: String, count: usize },
}

impl Finding<'_> {
    /// The finding's subject: its clause's name or its place's name.
    pub fn subject(&self) -> &str {
        match self {
            Finding::Clause { clause, .. } => &clause.name,
            Finding::Stranded { place, .. } => place,
        }
    }

    /// The finding in words.
    pub fn describe(&self) -> String {
        match self {
            Finding::Stranded { place, count } => {
                format!("{place} holds {count} at quiescence, and nothing in the contract lets a token rest there")
            }
            Finding::Clause { clause, count, bound } => {
                let expected = count_across(clause.min, clause.max, &clause.places);
                match bound {
                    CountBound::Upper => format!(
                        "{expected} at quiescence, found {count} (an upper bound holds under a terminal too)"
                    ),
                    CountBound::Lower => format!("{expected} at quiescence, found {count}"),
                }
            }
        }
    }
}

/// What the quiescent marking `m` breaks: clauses in contract order, then stranded places
/// by name. Empty when `m` meets the contract.
pub(super) fn quiescence_findings<'c>(
    m: &MarkingState,
    contract: &'c OpenNetContract,
    rest: &RestDeclaration,
) -> Vec<Finding<'c>> {
    let mut findings: Vec<Finding<'c>> = Vec::new();
    // The clause is a QuiescentCount with every terminal marker as a waiver ([VER-002]),
    // read through the same predicate the graph routes use for that property.
    let markers = waiver_markers(contract);
    for clause in contract.clauses() {
        if let Some(bound) = count_violation(m, &clause.places, clause.min, clause.max, &markers) {
            findings.push(Finding::Clause { clause, count: tokens_across(m, &clause.places), bound });
        }
    }
    for place in stranded_names(m, rest) {
        let count = m.count(&place);
        findings.push(Finding::Stranded { place, count });
    }
    findings
}
